encryption/src/main/java/org/apache/solr/encryption/EncryptionRequestHandler.java

```diff
@@ -334,75 +334,72 @@ public void handleRequestBody(SolrQueryRequest req, SolrQueryResponse rsp) throw
-      long timeMs = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startTimeNs);
+      rsp.addToLog(ENCRYPTION_STATE, state.value);
       log.info("Responding encryption state={} success={} for keyId={} timeMs={}",
-          state.value, success, keyId, timeMs);
+          state.value, success, keyId, elapsedTimeMs(startTimeNs));
     }
   }
 
   private void distributeRequest(SolrQueryRequest req, SolrQueryResponse rsp, String keyId, long startTimeNs) {
     boolean success = false;
-    String collectionName = null;
     State collectionState = null;
     long timeAllowedMs = req.getParams().getLong(TIME_ALLOWED, 0);
-    long maxTimeNs = timeAllowedMs <= 0 ? Long.MAX_VALUE : startTimeNs + timeAllowedMs;
+    TimeOut timeOut = timeAllowedMs <= 0 ? null : new TimeOut(timeAllowedMs, TimeUnit.MILLISECONDS, getTimeSource());
     try {
-      collectionName = req.getCore().getCoreDescriptor().getCollectionName();
+      String collectionName = req.getCore().getCoreDescriptor().getCollectionName();
       if (collectionName == null) {
         throw new SolrException(SolrException.ErrorCode.SERVER_ERROR, "Parameter " + DISTRIB + " can only be used in Solr Cloud mode.");
       }
-      log.debug("Encrypt request distributed for keyId={} collection={}", keyId, collectionName);
+      log.debug("Encrypt request distributed for keyId={}", keyId);
       DocCollection docCollection = req.getCore().getCoreContainer().getZkController().getZkStateReader().getCollection(collectionName);
       if (docCollection == null) {
         throw new SolrException(SolrException.ErrorCode.SERVER_ERROR, "Parameter " + DISTRIB + " present but collection '" + collectionName + "' not found.");
       }
-      try (SolrClientHolder solrClient = getHttpSolrClient(req)) {
-        ModifiableSolrParams params = createDistributedRequestParams(req, rsp, keyId);
-        Collection<Slice> slices = docCollection.getActiveSlices();
-        Collection<Callable<State>> encryptRequests = new ArrayList<>(slices.size());
-        final String collectionNameFinal = collectionName;
-        for (Slice slice : slices) {
-          Replica replica = slice.getLeader();
-          if (replica == null) {
-            log.error("No leader found for shard {}", slice.getName());
-            collectionState = State.ERROR;
-            continue;
-          }
-          encryptRequests.add(() -> sendEncryptionRequestWithRetry(replica, params, solrClient.getClient(), keyId, collectionNameFinal));
+      ModifiableSolrParams params = createDistributedRequestParams(req, rsp, keyId);
+      Collection<Slice> slices = docCollection.getActiveSlices();
+      Collection<Callable<State>> encryptRequests = new ArrayList<>(slices.size());
+      // Use the update-only http client, considering encryption is an update as we indeed create new Lucene segments.
+      Http2SolrClient solrClient = req.getCoreContainer().getUpdateShardHandler().getUpdateOnlyHttpClient();
```

Comment:
Seems you could move this to inside sendEncryptionRequestWithRetry, thus one fewer param as well.  It doesn't matter but sendEncryptionRequestWithRetry has lots of params so maybe this move would be nicer.  Up to you.